compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -1580,31 +1601,47 @@ impl<'tcx> LayoutCx<'tcx, TyCtxt<'tcx>> {
         // GeneratorLayout.
         debug!("prefix = {:#?}", prefix);
         let (outer_fields, promoted_offsets, promoted_memory_index) = match prefix.fields {
-            FieldsShape::Arbitrary { mut offsets, memory_index } => {
-                let mut inverse_memory_index = invert_mapping(&memory_index);
-
-                // "a" (`0..b_start`) and "b" (`b_start..`) correspond to
-                // "outer" and "promoted" fields respectively.
-                let b_start = (tag_index + 1) as u32;
-                let offsets_b = offsets.split_off(b_start as usize);
-                let offsets_a = offsets;
-
-                // Disentangle the "a" and "b" components of `inverse_memory_index`
-                // by preserving the order but keeping only one disjoint "half" each.
-                // FIXME(eddyb) build a better abstraction for permutations, if possible.
-                let inverse_memory_index_b: Vec<_> =
-                    inverse_memory_index.iter().filter_map(|&i| i.checked_sub(b_start)).collect();
-                inverse_memory_index.retain(|&i| i < b_start);
-                let inverse_memory_index_a = inverse_memory_index;
-
-                // Since `inverse_memory_index_{a,b}` each only refer to their
+            FieldsShape::Arbitrary { offsets, memory_index } => {
+                let inverse_memory_index = invert_mapping(&memory_index);
+
+                // Outer fields - upvars and tag
+                let after_tag = tag_index + 1;
+                let offsets_outer: Vec<_> = vec![offsets[upvar_count].clone()];
+                let mut inverse_memory_index_outer = inverse_memory_index.clone();
+                inverse_memory_index_outer.retain(|&i| i < after_tag as u32);
+
+                // Promoted fields - upvars and promoted locals
+                let offsets_promoted = offsets;
+                let inverse_memory_index_promoted = inverse_memory_index;
```

Comment:
I think we still need to decrease everything by 1

Don't remember why the code is like this, but anyway